accepted/2024/trusted-publishers-oidc-for-nuget-push.technical.md

```diff
@@ -0,0 +1,244 @@
+than 1% of additional coverage.
+
+<!--
+
+```kusto
+let MaxDistanceBetweenNeighbors = 5m;
+let MaxDistanceFromFirst = 30d;
+NiPackageVersions
+| where ResultType == "Available"
+| join kind=inner NiPackageOwners on LowerId
+| extend Owners = array_strcat(array_sort_asc(Owners), ", ")
+| order by Owners asc, Created asc
+| extend PushSession = row_window_session(Created, MaxDistanceFromFirst, MaxDistanceBetweenNeighbors, Owners != prev(Owners))
+| summarize PushCount = count(), PushSessionDuration = max(Created) - min(Created) by Owners, PushSession
+| where PushSession > ago(365d)
+| summarize (PushSessionDuration, PushCount) = arg_max(PushSessionDuration, PushCount) by Owners
+| summarize
+    ['% with single push'] = round(100.0 * countif(PushCount == 1) / count(), 2),
+    ['% < 1m'] = round(100.0 * countif(PushSessionDuration < 1m) / count(), 2),
+    ['% < 5m'] = round(100.0 * countif(PushSessionDuration < 5m) / count(), 2),
+    ['% < 10m'] = round(100.0 * countif(PushSessionDuration < 10m) / count(), 2),
+    ['% < 15m'] = round(100.0 * countif(PushSessionDuration < 15m) / count(), 2),
+    ['% < 30m'] = round(100.0 * countif(PushSessionDuration < 30m) / count(), 2),
+    ['% < 1h'] = round(100.0 * countif(PushSessionDuration < 1h) / count(), 2)
+```
+
+--> 
+
+| % with single push | % < 1m | % < 5m | % < 10m | % < 15m | % < 30m | % < 1h |
+| ------------------ | ------ | ------ | ------- | ------- | ------- | ------ |
+| 63.23              | 75.32  | 93.49  | 98      | 99.12   | 99.84   | 99.95  |
+
+## Other package sources
+
+Other NuGet package sources aside from NuGet.org could also implement this protocol. They would need to implement the
+token trade endpoint.
+
+The `nuget/login` action could be implemented so that it supports and V3 package source, as long as it has a
+`TokenService/1.0.0` resource in the service index. This level of flexibility should be implemented anyways so that it
+can be tested against NuGet DEV and INT pre-production environments.
+
+It would be the responsibility of the package source to implement OIDC token validation as well as expressing trust
+policies.
+
+## Auditing usage on NuGet.org
+
+To help us understand the success of this feature and record priviledged actions for security auditing, we will at least
+record minimal information about the OIDC token trade, such as the repository owner, repository name, workflow path,
+etc. Existing auditing for API key usage will be used anywhere the short-lived API key is used.
+
+The metadata available in a GitHub Actions OIDC token is mentioned in GitHub's [Understanding the OIDC
+token](https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect#understanding-the-oidc-token).
+
+These claims could offer useful indicators to package consumers about the package. In order to support a future effort
```

Comment:
This comment is orthogonal for trusted publishing, but I might avoid discussing SLSA Build L1 - It doesn't add much given it's not a signed attestation and if the registry was constructing the provenance, then it requires trust in the registry that it properly recorded the claims, rather than the builder.

Reply:
That's a fair point. I do want to mention SLSA Build levels to come extent, at least to say this does NOT provided L2+! I talked to Zach S. on the OpenSSF Slack. He said using this information would be L1 but it's not great since it is tamperable. I would say this metadata is better than package author-attested repo/project URL since the consumer needs to just trust the registry and GitHub Actions, but it's not as good as a proper signed provenance thing.